### ISSUE TITLE
Best way to remove an individual card?

### ISSUE THREAD

**Issue by theogravity** [trustly/pci-blackbox#1]
Hi there,

Thanks for taking the time to document and have code for your PCI process! I'm trying to make use of some of it myself and was wondering what is the best way to remove an individual card?

I noticed that the `CardNumberReference` field is not unique in the sense that you can have multiple copies of a card with the same `CardNumberReference` in the `encryptedcards` table. 

I feel that one would have to pass through the `cardkey` and regenerate the `CardKeyHash` value to remove the card in `encryptedcards`, but my concern is the possibility of a dangling reference in the `CardNumberReference` table

**Comment by joelonsql** [comment]
Yes, CardNumberReference will intentionally be the same between different users, which allows different users to compare which cards they have in common without sharing any card details with each other (not even hashes of card numbers as the CardNumberReference is a UUID).

If you want to remove a card for a single user, remove it from EncryptedCards, but if the card holder wishes to remove all cards for all users, remove it from both EncryptedCards and CardNumberReferences.

There is currently no functionality to remove cards via function calls.

**Comment by theogravity** [comment]
Thank you for your prompt reply! And really thank you for providing this framework. I'm only using the encrypt/decrypt card portions but it's helped a ton with understanding how this stuff should work.

Would be happy to make a donation if you guys are open to that!

**Comment by theogravity** [comment]
For others who are interested in deleting a card, I've made a function for it (open to code reviews)

```
-- Created by Theo Gravity

CREATE OR REPLACE FUNCTION Remove_Card(
    _CardKey text
) RETURNS boolean AS $BODY$
DECLARE
_CardKeyHash bytea;
BEGIN

_CardKeyHash := digest(_CardKey,'sha512');

DELETE FROM EncryptedCards WHERE cardkeyhash = _CardKeyHash;

-- see: http://www.postgresql.org/docs/current/static/plpgsql-statements.html#PLPGSQL-STATEMENTS-DIAGNOSTICS
RETURN FOUND;
END;
$BODY$ LANGUAGE plpgsql VOLATILE;
```
